src/Analyzers/Core/CodeFixes/SimplifyLinqExpression/AbstractSimplifyLinqExpressionCodeFixProvider`3.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.CodeAnalysis.SimplifyLinqExpression
+{
+    internal abstract class AbstractSimplifyLinqExpressionCodeFixProvider<TInvocationExpressionSyntax, TSimpleNameSyntax, TExpressionSyntax> : SyntaxEditorBasedCodeFixProvider
+        where TInvocationExpressionSyntax : SyntaxNode
+        where TSimpleNameSyntax : SyntaxNode
+        where TExpressionSyntax : SyntaxNode
+    {
+        private const string SimplyfyLinqAnnotationKind = "linqTracking";
+
+        protected abstract TSimpleNameSyntax GetName(TInvocationExpressionSyntax invocationExpression);
+        protected abstract TExpressionSyntax GetExpression(TExpressionSyntax invocationExpression);
+        protected abstract SyntaxNode[] GetArguments(SyntaxNode invocationExpression);
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds
+           => ImmutableArray.Create(IDEDiagnosticIds.SimplifyLinqExpressionDiagnosticId);
+
+        internal override CodeFixCategory CodeFixCategory => CodeFixCategory.CodeQuality;
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(new MyCodeAction(
+                c => FixAsync(context.Document, context.Diagnostics.First(), c)),
+                context.Diagnostics);
+            return Task.CompletedTask;
+        }
+
+        protected override Task FixAllAsync(Document document,
+                                            ImmutableArray<Diagnostic> diagnostics,
+                                            SyntaxEditor editor,
+                                            CancellationToken cancellationToken)
+        {
+            var root = editor.OriginalRoot;
+
+            // Track all the arguments into linq methods as they could be lambdas and the contents of their bodies may change
+            var invocationsAndTheirArguments = diagnostics.Select(d => (Invocation: GetInvocation(root, d), Arguments: GetArguments(root, d)));
+            var argumentLookup = invocationsAndTheirArguments
```

Comment:
this all seems surprisingly complex.  can you not just rewrite bottom up (order by span, descending), and then just do `editor.ReplaceNode(original, (current, _) => ProcessCurrrentHere(...)`

Reply:
that was my original approach. something like this: 
```cs
var nodes = diagnostics.Select(d => GetNodes(d, root));
var expressionsToReWrite = nodes .Select(x => x.Invocation).OrderByDescending(x => x.SpanStart);
var relatedNodesByInvocationSpanStart = nodes.ToImmutableDictionary(n => n.Invocation.SpanStart, n => (n.Expression, n.Name, n.ArgumentList));
foreach (var original in expressionsToReWrite)
{
    editor.ReplaceNode(original, (current, generator) =>
    {
        var (expression, name, argumentList) = relatedNodesByInvocationSpanStart[original.SpanStart];
        return generator.InvocationExpression(
                generator.MemberAccessExpression(expression, name),
                GetArguments(argumentList)).WithTriviaFrom(current);
    });
}
```


The problem I am having is that in this case:
```cs
var a = test.Where(a => a.Where(s => s.Equals(""hello"")).FirstOrDefault().Equals(""hello""))FirstOrDefault();
```

order-by-desceding means I replace:
```cs
a => a.Where(s => s.Equals(""hello"")).FirstOrDefault().Equals(""hello"")
```

with
```cs
a => a.FirstOrDefault(s => s.Equals(""hello"")).Equals(""hello"")
```
 and I need to re-find the new arguments in some way.

Otherwise the code above will just get me:
```cs
var a = test.FirstOrDefault(a => a.Where(s => s.Equals(""hello"")).FirstOrDefault().Equals(""hello""));
```

Reply:
instead of `var (expression, name, argumentList) = relatedNodesByInvocationSpanStart[original.SpanStart];` grab the data off of `current` :)

Reply:
I can try and re-write it. Today I look nodes up via the spans from the diagnostics. The nodes in `current` will not have the same spans as the original that we did the analysis for.

Reply:
i can help with this i think :)  we can discuss offline, or i can see if i can make some changes  here.